File: src/main/kotlin/baekjun/binary-search/1654-랜선 자르기/1.kt
package algo

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*
import kotlin.math.max

// 각 랜선 길이 2^31 -1 이하
// 총 랜선 개수 <= 10_000
// 필요한 랜선 개수 <= 1_000_000 -> Long 가능
val br = BufferedReader(InputStreamReader(System.`in`))
val bw = BufferedWriter(OutputStreamWriter(System.out))
val arr = LongArray(10_000)
var ans = 0L
fun main() {
    val (k, n) = br.readLine().split(" ").map { it.toInt() }
    repeat(k) { i ->
        arr[i] = br.readLine().toLong()
    }
    val end = arr.maxOrNull() ?: 0
    binarySearch(1, end, n)
    bw.write("$ans")
    bw.flush()
    bw.close()
    br.close()
}

private fun binarySearch(start: Long, end: Long, n: Int) {
    if (end < start) return
    val mid = (start + end) / 2
    val cnt = count(mid)
    if (cnt < n) {
        binarySearch(start, mid - 1, n)
    } else {
        ans = max(ans, mid) // cnt 가 baekjun.getN 이상 이기만 하면 되니까
        binarySearch(mid + 1, end, n)
    }
}

private fun count(x: Long): Long {
    var result = 0L
    arr.forEach { it ->
        result += it / x
    }
    return result
}

File: src/main/kotlin/baekjun/binary-search/1654-랜선 자르기/1.py
k, baekjun.getN = map(int,input().split())
# k : 주어진 랜선 수
# baekjun.getN : 필요한 랜선 수

arr = [int(input())for _ in range(k)]
# arr : 주어진 랜선의 길이 정보 list

# 조건
# N개보다 많이 만드는 것도 N개를 만드는 것에 포함된다.
# k: 1~10,000 , baekjun.getN: 1~100만, 랜선길이 1~2^31-1, k<=baekjun.getN

# 풀이

# x 길이로 나누었을 때 얻는 랜선 개수

def count(arr,x):
    result = 0
    for i in arr:
        result += i//x

    return result

    # print(count(arr,200)) # 11

# 이분 탐색: 랜선 길이

start = 1
end = max(arr)
ans_list = []
# baekjun.getN : 필요한 랜선 수
def binary_search(arr,start,end):

    if start > end:

        return

    mid = (start + end) //2

    if count(arr,mid) < baekjun.getN : # mid가 작아지면 count 수가 커진다.
        binary_search(arr,start, mid-1)

    else:
        ans_list.append(mid)
        binary_search(arr,mid+1, end)



binary_search(arr,start,end)

print(max(ans_list))
